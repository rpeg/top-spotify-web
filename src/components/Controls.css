.controls {
    display:table;
    list-style:none;
    padding-left:0;
  }
  
.controls > li 
{
    display:table-row;
}

.controls > li > *
{
    display:table-cell;
    padding:5px;
}

.controls label {
    font-weight:bold;
    text-align:right;
}

.controls label:after {
    content: ':';
}