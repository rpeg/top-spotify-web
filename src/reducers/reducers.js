import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
