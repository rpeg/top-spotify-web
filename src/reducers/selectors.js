export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
