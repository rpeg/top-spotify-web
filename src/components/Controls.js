import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  Button, Container, Col, Row,
} from 'react-bootstrap';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import Slider from '@material-ui/core/Slider';
import Switch from '@material-ui/core/Switch';
import MenuItem from '@material-ui/core/MenuItem';
import { makeStyles } from '@material-ui/core/styles';
import domtoimage from 'dom-to-image';

import {
  TimeRanges,
} from '../constants/constants';
import {
  setTimeRangeName,
  setArtistCount,
  setTrackCount,
  setGenreCount,
  setHasClickedCreate,
  setStatsOptions,
  setDisplayProfile,
  setOptimizeTracks,
} from '../actions/actions';
import * as selectors from '../reducers/selectors'

const useStyles = makeStyles((theme) => ({
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
  label: {
    '&.MuiFormLabel-root': {
      color: '#FFFFFF',
      fontFamily: "'Karla', sans-serif",
    },
  },
  select: {
    maxWidth: 120,
    color: '#FFFFFF',
    fontFamily: "'Karla', sans-serif",

    '&.MuiInput-underline:before': {
      borderBottom: '1px solid rgba(255, 255, 255, 0.42)',
    },
    '& svg': {
      color: '#FFFFFF',
    },
  },
  slider: {
    '&.MuiSlider-root': {
      color: '#1DB954',
      marginTop: '25px',
    },
  },
  switch: {
    '&.MuiSwitch-root': {
      marginTop: '15px',
    },
    '& .MuiSwitch-colorSecondary.Mui-checked': {
      color: '#1DB954',
    },
    '& .MuiSwitch-colorSecondary.Mui-checked + .MuiSwitch-track': {
      backgroundColor: '#1DB954',
    },
  },
}));

const Controls = () => {
  const user = useSelector(selectors.selectUser);
  const timeRangeName = useSelector(selectors.selectTimeRangeName);
  const genreCount = useSelector(selectors.selectGenreCount);
  const artistCount = useSelector(selectors.selectArtistCount);
  const trackCount = useSelector(selectors.selectTrackCount);
  const statsOptions = useSelector(selectors.selectStatsOptions);
  const hasClickedCreate = useSelector(selectors.selectHasClickedCreate);
  const displayProfile = useSelector(selectors.selectDisplayProfile);
  const optimizeTracks = useSelector(selectors.selectOptimizeTracks);

  const [selectedTimeRangeName, setSelectedTimeRangeName] = useState(timeRangeName);
  const [selectedArtistCount, setSelectedArtistCount] = useState(artistCount);
  const [selectedTrackCount, setSelectedTrackCount] = useState(trackCount);
  const [selectedGenreCount, setSelectedGenreCount] = useState(genreCount);
  const [selectedStatsOptions, setSelectedStatsOptions] = useState(statsOptions);
  const [selectedDisplayProfile, setSelectedDisplayProfile] = useState(displayProfile);
  const [selectedOptimizeTracks, setSelectedOptimizeTracks] = useState(optimizeTracks);

  const dispatch = useDispatch();

  const processClick = () => {
    if (selectedTimeRangeName !== timeRangeName) {
      dispatch(setTimeRangeName(selectedTimeRangeName));
    }
    if (selectedGenreCount !== genreCount) {
      dispatch(setGenreCount(selectedGenreCount));
    }
    if (selectedArtistCount !== artistCount) {
      dispatch(setArtistCount(selectedArtistCount));
    }
    if (selectedTrackCount !== trackCount) {
      dispatch(setTrackCount(selectedTrackCount));
    }
    if (selectedStatsOptions !== statsOptions) {
      dispatch(setStatsOptions(selectedStatsOptions));
    }
    if (selectedDisplayProfile !== displayProfile) {
      dispatch(setDisplayProfile(selectedDisplayProfile));
    }
    if (selectedOptimizeTracks !== optimizeTracks) {
      dispatch(setOptimizeTracks(selectedOptimizeTracks));
    }

    dispatch(setHasClickedCreate());
  };

  const classes = useStyles();

  const handleSave = () => {
    const scale = 2;
    const elm = document.getElementById('top-spotify');

    // Transform hack to minimize rasterization artifacts
    domtoimage.toPng(elm, {
      height: elm.offsetHeight * scale,
      style: {
        transform: `scale(${scale}) translate(${elm.offsetWidth / 2 / scale}px, ${elm.offsetHeight / 2 / scale}px)`,
      },
      width: elm.offsetWidth * scale,
    })
      .then((dataUrl) => {
        window.open().document.write(`<div><img src="${dataUrl}" style="position: absolute; width: 100%"/></div>`);
      })
      .catch((err) => {
        console.error('Could not export PNG', err);
      });
  };

  return (
    user && user.id ? (
      <div style={{ borderBottom: '1px solid white' }}>
        <Container style={{ paddingTop: '30px' }}>
          <Row className="justify-content-center">
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  Period
                </InputLabel>
                <Select
                  value={selectedTimeRangeName}
                  className={classes.select}
                  onChange={(e) => setSelectedTimeRangeName(e.target.value)}
                >
                  <MenuItem value={TimeRanges.LONG.name}>{TimeRanges.LONG.title}</MenuItem>
                  <MenuItem value={TimeRanges.MEDIUM.name}>{TimeRanges.MEDIUM.title}</MenuItem>
                  <MenuItem value={TimeRanges.SHORT.name}>{TimeRanges.SHORT.title}</MenuItem>
                </Select>
              </FormControl>
            </Col>
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  # of genres
                </InputLabel>
                <Slider
                  defaultValue={selectedGenreCount}
                  step={1}
                  valueLabelDisplay="auto"
                  min={0}
                  max={100}
                  className={classes.slider}
                  onChange={(e, newValue) => setSelectedGenreCount(newValue)}
                />
              </FormControl>
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  # of artists
                </InputLabel>
                <Slider
                  defaultValue={selectedArtistCount}
                  step={5}
                  marks
                  valueLabelDisplay="auto"
                  min={0}
                  max={50}
                  className={classes.slider}
                  onChange={(e, newValue) => setSelectedArtistCount(newValue)}
                />
              </FormControl>
            </Col>
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  # of tracks
                </InputLabel>
                <Slider
                  defaultValue={selectedTrackCount}
                  step={5}
                  marks
                  valueLabelDisplay="auto"
                  min={0}
                  max={50}
                  className={classes.slider}
                  onChange={(e, newValue) => setSelectedTrackCount(newValue)}
                />
              </FormControl>
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  Stats to display
                </InputLabel>
                <Select
                  value={selectedStatsOptions}
                  multiple
                  className={classes.select}
                  onChange={(e) => setSelectedStatsOptions(e.target.value)}
                >
                  <MenuItem value="key">Key</MenuItem>
                  <MenuItem value="bpm">BPM</MenuItem>
                  <MenuItem value="decades">Decades</MenuItem>
                  <MenuItem value="scale">Scale</MenuItem>
                  <MenuItem value="features">Features</MenuItem>
                </Select>
              </FormControl>
            </Col>
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  Optimize tracks
                </InputLabel>
                <Switch
                  className={classes.switch}
                  checked={selectedOptimizeTracks}
                  onChange={(e) => setSelectedOptimizeTracks(e.target.checked)}
                />
              </FormControl>
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col xs={4}>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="demo-simple-select-placeholder-label-label" className={classes.label}>
                  Display profile
                </InputLabel>
                <Switch
                  className={classes.switch}
                  checked={selectedDisplayProfile}
                  onChange={(e) => setSelectedDisplayProfile(e.target.checked)}
                />
              </FormControl>
            </Col>
            <Col xs={4} />
          </Row>
          <Row className="justify-content-center">
            <Col xs={3}>
              <Button
                variant="outline-primary"
                style={{ margin: '30px 0 30px 0' }}
                onClick={processClick}
              >
                Create
              </Button>
            </Col>
            <Col xs={3}>
              <Button
                variant="outline-primary"
                style={{ margin: '30px 0 30px 0' }}
                onClick={handleSave}
                disabled={!hasClickedCreate}
              >
                Export
              </Button>
            </Col>
          </Row>
        </Container>
      </div>
    ) : null
  );
};

export default Controls;
