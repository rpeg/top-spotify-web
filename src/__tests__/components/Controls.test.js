import React from 'react';
import { shallow } from 'enzyme';

import Controls from '../../components/Controls';
import * as selectors from '../../reducers/selectors';
import {
  TimeRanges, DEFAULT_GENRE_COUNT, DEFAULT_ARTIST_COUNT, DEFAULT_TRACK_COUNT, DEFAULT_STATS_OPTIONS,
} from '../../constants/constants';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = (
  user = {},
  timeRangeName = TimeRanges.LONG.name,
  genreCount = DEFAULT_GENRE_COUNT,
  artistCount = DEFAULT_ARTIST_COUNT,
  trackCount = DEFAULT_TRACK_COUNT,
  statsOptions = DEFAULT_STATS_OPTIONS,
  hasClickedCreate = false,
  displayProfile = true,
  optimizeTracks = true,
) => {
  jest.spyOn(selectors, 'selectUser').mockReturnValue(user);
  jest.spyOn(selectors, 'selectTimeRangeName').mockReturnValue(timeRangeName);
  jest.spyOn(selectors, 'selectGenreCount').mockReturnValue(genreCount);
  jest.spyOn(selectors, 'selectArtistCount').mockReturnValue(artistCount);
  jest.spyOn(selectors, 'selectTrackCount').mockReturnValue(trackCount);
  jest.spyOn(selectors, 'selectStatsOptions').mockReturnValue(statsOptions);
  jest.spyOn(selectors, 'selectHasClickedCreate').mockReturnValue(hasClickedCreate);
  jest.spyOn(selectors, 'selectDisplayProfile').mockReturnValue(displayProfile);
  jest.spyOn(selectors, 'selectOptimizeTracks').mockReturnValue(optimizeTracks);
};

describe('Controls', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    setup();
    const wrapper = shallow(<Controls />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    setup();
    const wrapper = shallow(<Controls />);
    expect(wrapper).toMatchSnapshot();
  });
});
