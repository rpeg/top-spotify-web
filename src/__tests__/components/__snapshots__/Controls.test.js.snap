// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders correctly 1`] = `""`;
